src/graph/LdapAuthenticator.cpp

```diff
@@ -0,0 +1,290 @@
+        return Status::Error("LDAP authentication cannot use ldap_searchattribute "
+                             "together with ldap_searchfilter.");
+    }
+
+    return Status::OK();
+}
+
+
+Status initLDAPConnection() {
+    int ldap_version = LDAP_VERSION3;
+
+    int ret  = ldap_create(&ldap_);
+    if (ret != LDAP_SUCCESS) {
+        return Status::Error("Init LDAP failed.");
+    }
+    ldap_->ld_options.ldo_defport = FLAGS_ldap_port;
+
+    ret = ldap_set_option(ldap_, LDAP_OPT_HOST_NAME, FLAGS_ldap_server);
+    if (ret != LDAP_SUCCESS) {
+        ldap_ld_free(ldsp_, 1, NULL, NULL);
+        return Status::Error("Init LDAP failed.");
+    }
+
+    ret = ldap_set_option(ldap_, LDAP_OPT_PROTOCOL_VERSION, &ldap_version);
+    if (ret != LDAP_SUCCESS) {
+        ldap_unbind(ldap_);
+        return Status::Error("Set LDAP protocol version failed");
+    }
+
+    if (FLAGS_ldap_tls) {
+        ret = ldap_start_tls_s(ldap_, NULL, NULL));
+        if (ret != LDAP_SUCCESS) {
+            ldap_unbind(ldap_);
+            return Status::Error("Start LDAP TLS session failed");
+        }
+    }
+
+    return Status::OK();
+}
+
+
+std::string LdapAuthenticator::buildFilter() {
+    std::string filter;
+    if (!FLAGS_ldap_searchfilter.empty()) {
+        filter = buildSearchFilter();
+    } else if (!FLAGS_ldap_searchattribute.empty()) {
+        filter = folly::stringPrintf("(%s=%s)", FLAGS_ldap_searchattribute.c_str(),
+                                      userName_.c_str());
+    } else {
+        filter = folly::stringPrintf("(uid=%s)", userName_.c_str());
+    }
+    return filter;
+}
+
+
+std::string LdapAuthenticator::buildSearchFilter() {
+    std::string result  = FLAGS_ldap_searchfilter;
+    std::string ph = "$username";
+    auto len = ph.length();
+
+    std::string::size_type  pos = 0;
+    while ((pos = result.find(ph)) != std::string::npos) {
+        result.replace(pos, len, userName_);
+    }
+
+    return result;
+}
+
+
+Status<bool> LdapAuthenticator::searchBindAuth() {
+    LDAPMessage *searchMessage;
+    LDAPMessage *entry;
+    char* distName;
+    std::string dn;
+
+    /**
+     * Firstly, perform the LDAP seach to find the distinguished name
+     * of the user who is logging in.
+     */
+
+    /**
+     * LDAP authentication disallows any characters that need to escape
+     * in user name.
+     */
+    for (auto& c : user) {
+        if (c == '*' ||
+            c == '(' ||
+            c == ')' ||
+            c == '/' ||
+            c == '\\') {
+            ldap_unbind(ldap_);
+            return Status::Error("User name contains invalid character in LDAP authentication.");
+        }
+    }
+
+    /**
+     * Bind with a pre-defined user name and password to search.
+     * If none is specified, use the anonymous bind.
+     */
+    auto& binddn = FLAGS_ldap_binddn;
+    auto& bindPassword = FLAGS_ldap_bindPassword;
+
+    auto ret = ldap_simple_bind_s(ldap_, binddn.c_str(), bindPassword.c_str());
```

Comment:
`ldap_simple_bind_s` and other interfaces here are all synchronous, right? If so, do the LDAP kits provide async ones?
We cannot do such blocking actions directly in the main working threads.

Reply:
Good point.  I will research it.  
![image](https://user-images.githubusercontent.com/59197347/75644535-9513ed80-5c7d-11ea-9f61-6fb232d4e9c4.png)

Reply:
@dutor 
When a user logs in from the console, graphd will start a thread to process.
Here is login verification, asynchronous synchronization doesn't feel that different.
In synchronization, function A connects to the LDAP server. Before the LDAP server returns, A is blocked.
Asynchronously, function B connects to the LDAP server. Before the LDAP server returns, B can continue to execute, but it will block behind B and wait for the result from the LDAP server.
So the difference between synchronous and asynchronous is not obvious